### PR TITLE
Add cancel() method to streams

### PR DESCRIPTION
**Pull request by voxpelli**
I'm unfortunately somewhat rusty in this regard, but I think this adds the possibility of canceling an infinite response as requested by me in #203 (and maybe as requested in #95).

This would add a `cancel()` that at least somewhat mimics the browser's `ReadableStream.cancel()`. To fully mimic it would probably have to [return a Promise](https://streams.spec.whatwg.org/#rs-cancel), but as there's no callback to the `message.destroy()` in node.js it would have to just be a `Promise.resolve()` then.

Thoughts on this approach? Good or bad?

Quickly looking at v2 it seems like this could be easily forward ported to that version.

### CODE REVIEW

**Comment by jimmywarting** [comment]
Actually... `cancel()` is something that is part of the ReadableStream api not fetch per say...
The `cancel` method should actually do a little bit more then simply call `res.destroy()` It should only cancel the stream if there is nothing else that is consuming the stream (aka if it's locked to something else).

The stream gets consumed/locked when you call `.blob()`,  `.arrayBuffer()`,  `.text()` or `.body.getReader()`

```js
var body = new Response('abc').body
var reader = body.getReader()
console.log(body.locked) // true
// The reason i think it return's a promise is cuz it can throw a error (Perhaps they thought it's nicer if it silence reject then throwing and making the rest crash)
body.cancel() // Uncaught TypeError: ReadableStreamReader constructor can only accept readable streams that are not yet locked to a reader
```

What needs to be done is 
```js
var body = new Response('abc').body
var reader = body.getReader()
reader.releaseLock() // Release the lock so no stream are using it
console.log(body.locked) // false
body.cancel() // now it will cancel
```

But the ReadableStream API are different then the Node stream Api
I'm not sure where we stand on implementing ReadableStream api in v2... The hole node eco system is built around node-streams so what would be the benefit of having ReadableStream api if you can't pipe it to a file?

**Comment by voxpelli** [comment]
@jimmywarting I'm mostly just trying to figure out a solution to #203, this might not be it, but the ability to cancel a never-ending infinite response is something I need no matter what kind of streams model it would be modeled around. Suggestion on how to do this with node streams?

**Comment by jimmywarting** [comment]
Yea, I agree, something is needed. We may need some differentiation in node-fetch since it's built on node-stream

i don't know if it's possible to detect if a stream has been locked to something
if you where to example do

```js
body = new Response('abc').body
body.pipe(new fs.createWriteStream('cache.bin'))
// Do we know if there is something that is consuming the stream?
// ideally `body.locked` should return true and therefore cancel should not work
```

I would be fine with a less stricter api and allow canceling a response even if that means something else is consuming the stream. After all it's two different streaming api

**Comment by bitinn** [comment]
I might be wrong as I haven't used this personally.

- The original `cancel()` is really intended to be used when you don't care about the result *at all*. And as far as I know, this won't be added to spec.
- The new approach appear to be *cancel token*, see [proposal-cancelable-promises](https://github.com/tc39/proposal-cancelable-promises) and an [implementation](https://github.com/mzabriskie/axios/pull/452) from [axios](https://github.com/mzabriskie/axios#cancellation), though it was not clear [how they handle streaming and cancelling](https://github.com/mzabriskie/axios/issues/499).

So to sum up the current status quo:

- If you want to abort and discard the body, *cancel token* can help.
- But if you want to stop the stream somehow, this is really a [Node.js problem](http://stackoverflow.com/questions/19277094/how-to-close-a-readable-stream-before-end).

I don't think we should add `res.body.cancel()` because it confuses people, but if the case for real-time JSON streaming is really popular, we might consider adding an extension to allow developers to control termination of underlying http request...

**Comment by TimothyGu** [comment]
The `cancel()` is already in the spec as part of the [`ReadableStream`](https://streams.spec.whatwg.org/#rs-cancel) interface. However I'm not convinced we should augment a Node.js `ReadableStream` object with WHATWG `ReadableStream` semantics -- doing so might trick the user into thinking the `res.body` we provide is actually a WHATWG `ReadableStream`, which is not the case.

I wouldn't mind adding support for Promise-based cancelation, but I don't think we should have our own implementation of `CancelToken` either. Nor is there any standalone polyfill for `CancelToken` available on npm AFAICT.

I'd agree with exposing the original `ClientRequest` object. I think it can be beneficial to Node.js users in general to have access to that object.

**Comment by bitinn** [comment]
@TimothyGu I haven't tried this, but does calling `destroy()` on a custom `agent` passed to `node-fetch` works? If so, we might not need to expose ClientRequest (I think 99% use case of Fetch is to avoid dealing with http.request directly?).

**Comment by voxpelli** [comment]
I'm aware of cancelable promises, so just to be clear: This and #203 is _not_ about cancelable _requests_, it's about fully resolved requests where one wants to _cancel the response_.

An example of a properly fulfilled request where the response needs to be canceled:

```javascript
fetch('http://example.com/infinite-response').then(res => {
  // The request is done, but now I want to abort the response
});
```

This type of streaming through an infinite responses is rather common and was in the form of [newline delimited JSON](https://en.wikipedia.org/wiki/JSON_Streaming#Line_delimited_JSON) popularised by Twitter's [streaming API](https://dev.twitter.com/streaming/overview/processing) and is made easy in Node.js through packages like [ndjson](https://www.npmjs.com/package/ndjson).

It all works great with node-fetch – apart from the fact that the response  can't ever be destroyed.

Abusing HTTP agents to ensure that the response is destroyed feels wrong to me. HTTP agents are responsible for many other things, including keep-alive, which may be useful for also this type of request.

It's also not enough to just stop the stream – the response also needs to be destroyed so that the http connection is released or destroyed.

The simplest way of ensuring that the response is destroyed would be to let the one who knows that the request should be destroyed also be the one who destroys it.

Therefore a `.cancel()` (perhaps with another name) would probably be the easiest solution to this problem.

**Comment by bitinn** [comment]
@voxpelli

I believe status quo is:

- WHATWG stream `cancel()` can't be used when you are actively piping to something.
- NODE stream doesn't have a `cancel()`, however HTTP request do have `destroy()` and `abort()`.

I don't think `res.destroy()` is more clean than `agent.destroy()`. And by that I mean they both forcefully close the socket, see [`socket.destroy() doc`](https://nodejs.org/api/net.html#net_socket_destroy_exception).

[`request.abort()`](https://nodejs.org/api/http.html#http_request_abort) is nicer, it just throw away any remaining data and then close socket. Not sure if it works for your case though.

@TimothyGu 

> I'd agree with exposing the original ClientRequest object. I think it can be beneficial to Node.js users in general to have access to that object.

How do you propose we expose this object? Through [event like got](https://github.com/sindresorhus/got#onrequest-request)?

**Comment by TimothyGu** [comment]
What I had on my mind was as a property on the `Response` object:

```js
fetch('...').then(res => res.clientRequest.cancel());
```

> it just throw away any remaining data and then close socket. Not sure if it works for your case though.

Don't see a reason why it wouldn't...

**Comment by voxpelli** [comment]
The suggestion of @TimothyGu would work for me, I updated the PR to match that with the one exception that I named the property `clientResponse` as it's a response object rather than the request object that's exposed.

This way one can do:

```javascript
fetch('...').then(res => res.clientResponse.destroy());
```

File: index.js [hunk 1]
```diff
@@ -176,6 +176,9 @@ function Fetch(url, opts) {
 
 			// prepare response
 			var body = res.pipe(new stream.PassThrough());
+
+			body.clientResponse = res;
```
Comment (TimothyGu) [inline index.js]: When decompression is requested, `body` is overwritten, so this `clientResponse` will get lost.

I was thinking of passing `req` as `clientRequest` through `response_options` into the `Response` constructor, and make `clientRequest` a property of the `Response` class. This way, not only would one be able to do:

```js
fetch('...').then(res => res.clientRequest.abort());
```

and cancel both the request and the response, which is especially helpful if chunked encoding is used.

**Comment by heri16** [comment]
@TimothyGu this was closed without being merged?
